project/SPIHelpers.scala

```diff
@@ -0,0 +1,72 @@
+import sbt.*
+import sbt.Keys.*
+
+object SPIHelpers {
+
+  /** A helper task that ensures consistency of SPI services definitions.
+    *
+    * It should be attached as a dependency to `Compile / compile / compileInputs` of a given library.
+    *
+    * It detects any unknown classes in the `services` definitions and forces a recompilation if needed, to ensure the
+    * consistency of the definitions. Without this helper task, the incremental compiler did not detect removed service
+    * classes, thus after such a class was removed, it still stayed in the SPI configuration - crashing at runtime when
+    * the missing class was attempted to be instantiated. Only a `clean` allowed to regenerate the SPI configuration.
+    * This was causing issues on the CI when switching between PRs that have some new SPI configurations - they were
+    * leaking and crashing unrelated PRs.
+    *
+    * This task is created with the `std-*` Java helper libraries in mind and is aimed primarily at Java-only projects.
+    * Additional tweaks may be needed to get it working for mixed Java/Scala projects, if ever needed.
+    *
+    * @see https://github.com/enso-org/enso/pull/8129
+    */
+  def ensureSPIConsistency = Def.task {
+    val log            = streams.value.log
+    val classDir       = (Compile / compile / classDirectory).value
+    val javaSourcesDir = (Compile / compile / javaSource).value
+    val serviceDir     = classDir / "META-INF" / "services"
+    log.debug(s"Scanning $serviceDir for SPI definitions.")
+
+    val files: Array[File] =
+      if (serviceDir.exists()) IO.listFiles(serviceDir) else Array()
+
+    files.foreach { serviceConfig =>
+      log.debug(s"Processing service definitions: $serviceConfig")
+      val definedClasses =
+        IO.readLines(serviceConfig).map { qualifiedClassName =>
+          val subPath        = qualifiedClassName.replace('.', '/')
```

Comment:
Won't work on windows?

Reply:
I tested it on Windows 😄 

Java does the path translation correctly, i.e. you can specify a `/` path and it is resolved correctly on Windows. Never had trouble with it.